---
title: "DESeq"
author: "Gwen"
date: "2023-10-18"
output: html_document
---

libraries needed
```{r}
library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(wesanderson)
library(vsn)
library(pheatmap)
library(eulerr)
```


Import Counts Matrix and Sample ID tables into R and DESeq2
```{r}

# Import the counts matrix
countsTable <- read.table("salmon.isoform.counts.matrix.filteredAssembly", header=TRUE, row.names=1)
head(countsTable)
dim(countsTable)

countsTableRound <- round(countsTable) # bc DESeq2 doesn't like decimals (and Salmon outputs data with decimals)
head(countsTableRound)

#import the sample discription table
conds <- read.delim("ahud_samples_R.txt", header=TRUE, stringsAsFactors = TRUE, row.names=1)
head(conds)
```


Explore the counts data a bit
```{r}
# Let's see how many reads we have for each sample
colSums(countsTableRound)
mean(colSums(countsTableRound)) # 14,687,388

barplot(colSums(countsTableRound), names.arg=colnames(countsTableRound),cex.names=0.5, las=3,ylim=c(0,21000000))
abline(h=mean(colSums(countsTableRound)), col="blue", lwd=2)

# the average number of counts per gene
rowSums(countsTableRound)
mean(rowSums(countsTableRound)) # 8217.81 (new assembly may have a lower mapping rate than the original one but there are fewer genes to map to so these reads get collapsed and the number of counts per gene increases)
median(rowSums(countsTableRound)) # 377

apply(countsTableRound,2,mean) # 2 in the apply function does the action across columns
apply(countsTableRound,1,mean) # 1 in the apply function does the action across rows
hist(apply(countsTableRound,1,mean),xlim=c(0,10000), ylim=c(0,60000),breaks=1000) # lots of genes with low expression

```


Create a DESeq object and define the experimental design
```{r}
#### Create a DESeq object and define the experimental design here with the tilde
dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData=conds, 
                              design= ~ generation + treatment)

dim(dds)

# Filter out genes with too few reads - remove all genes with counts < 15 in more than 75% of samples, so ~28 (n=38)
## suggested by WGCNA on RNAseq FAQ
dds <- dds[rowSums(counts(dds) >= 15) >= 28,]
nrow(dds) # [1] 41348 (30 reads) [2] 25260 (new assembly, 15 reads)

# Run the DESeq model to test for differential gene expression
dds <- DESeq(dds)

# List the results you've generated
resultsNames(dds)
```

Check the quality of the data by sample clustering and visualization
```{r}
# The goal of transformation "is to remove the dependence of the variance on the mean, particularly the high variance of the logarithm of count data when the mean is low."

library(pheatmap)
library(vsn)

# this gives log2(n + 1)
ntd <- normTransform(dds)
meanSdPlot(assay(ntd)) # red fitted line shows that standard deviation for transcripts is pretty low (rank(mean)=expression level)

# Variance stabilizing transformation
vsd <- vst(dds, blind=FALSE) # reduces variation in more lowly expressed transcripts
meanSdPlot(assay(vsd))
           

sampleDists <- dist(t(assay(vsd)))

library(RColorBrewer)
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$treatment, vsd$generation, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

PCA to visualize global gene expression patterns
```{r}
# first transform the data for plotting using variance stabilization
vsd <- vst(dds, blind=FALSE)

pcaData <- plotPCA(vsd, intgroup=c("treatment","generation"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"))

ggplot(pcaData, aes(PC1, PC2, color=treatment, shape=generation)) +.  #more clustering by generation than by treatment
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() 
```

Let's plot the PCA by generation in four panels
```{r}
data <- plotPCA(vsd, intgroup=c("treatment","generation"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"))

###########  

dataF0 <- subset(pcaData, generation == 'F0')

F0 <- ggplot(dataF0, aes(PC1, PC2)) +
  geom_point(size=10, stroke = 1.5, aes(fill=treatment, shape=treatment)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  ylim(-10, 25) + xlim(-40, 10)+ # zoom for F0 with new assembly
  #ylim(-40, 25) + xlim(-50, 50)+ # new assembly limits
  #ylim(-40, 20) + xlim(-50, 30)+
  scale_shape_manual(values=c(21,22,23,24), labels = c("Ambient", "Acidification","Warming", "OWA"))+
  scale_fill_manual(values=c('#6699CC',"#F2AD00","#00A08A", "#CC3333"), labels = c("Ambient", "Acidification","Warming", "OWA"))+
  ##theme(legend.position = c(0.83,0.85), legend.background = element_blank(), legend.box.background = element_rect(colour = "black")) +
  #guides(shape = guide_legend(override.aes = list(shape = c( 21,22, 23, 24))))+
  #guides(fill = guide_legend(override.aes = list(shape = c( 21,22, 23, 24))))+
  #guides(shape = guide_legend(override.aes = list(size = 5)))+
  theme_bw() +
  theme(legend.position = "none") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 4))+
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank())

F0


#png("PCA_F0.png", res=300, height=5, width=5, units="in")
#ggarrange(F0, nrow = 1, ncol=1)
#dev.off()

################# F2

dataF2 <- subset(pcaData, generation == 'F2')

F2 <- ggplot(dataF2, aes(PC1, PC2)) +
  geom_point(size=10, stroke = 1.5, aes(fill=treatment, shape=treatment)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  ylim(-40, 25) + xlim(-50, 55)+
  #ylim(-40, 20) + xlim(-50, 30)+
  scale_shape_manual(values=c(21,22,23), labels = c("Ambient", "Acidification","Warming"))+
  # scale_color_manual(values = c('#6699CC',"#F2AD00","#00A08A", "#CC3333")) + 
  #scale_color_manual(values=c('black')) +
  scale_fill_manual(values=c('#6699CC',"#F2AD00","#00A08A"), labels = c("Ambient", "Acidification","Warming"))+
  theme(legend.position = c(0.83,0.85), legend.background = element_blank(), legend.box.background = element_rect(colour = "black")) +
  #scale_size(guide="none") +
  guides(shape = guide_legend(override.aes = list(shape = c( 21,22, 23))))+
  guides(fill = guide_legend(override.aes = list(shape = c( 21,22, 23))))+
  guides(shape = guide_legend(override.aes = list(size = 5)))+
  theme_bw() +
  theme(legend.position = "none") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 4))+
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank())
F2


# png("PCA_F2.png", res=300, height=5, width=5, units="in")
# ggarrange(F2, nrow = 1, ncol=1)
# dev.off()

# Yes - F2 is missing one ambient replicate

################################ F4

dataF4 <- subset(pcaData, generation == 'F4')

F4 <- ggplot(dataF4, aes(PC1, PC2)) +
  geom_point(size=10, stroke = 1.5, aes(fill=treatment, shape=treatment)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  ylim(-40, 25) + xlim(-50, 55)+ # limits with filtered assembly
  #ylim(-20, 10) + xlim(-40, 25)+  # zoom with filtered assembly
  #ylim(-40, 20) + xlim(-50, 30)+
  scale_shape_manual(values=c(21,22,23,24), labels = c("Ambient", "Acidification","Warming", "OWA"))+
  # scale_color_manual(values = c('#6699CC',"#F2AD00","#00A08A", "#CC3333")) + 
  #scale_color_manual(values=c('black')) +
  scale_fill_manual(values=c('#6699CC',"#F2AD00","#00A08A", "#CC3333"), labels = c("Ambient", "Acidification","Warming", "OWA"))+
  #theme(legend.position = c(0.83,0.85), legend.background = element_blank(), legend.box.background = element_rect(colour = "black")) +
  #scale_size(guide="none") +
  guides(shape = guide_legend(override.aes = list(shape = c( 21,22, 23, 24))))+
  guides(fill = guide_legend(override.aes = list(shape = c( 21,22, 23, 24))))+
  guides(shape = guide_legend(override.aes = list(size = 5)))+
  theme_bw() +
  theme(legend.position = "none") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 4))+
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank())
F4


# png("PCA_F4.png", res=300, height=5, width=5, units="in")
# ggarrange(F4, nrow = 1, ncol=1)
# dev.off()


################# F11

dataF11 <- subset(pcaData, generation == 'F11')

F11 <- ggplot(dataF11, aes(PC1, PC2)) +
  geom_point(size=10, stroke = 1.5, aes(fill=treatment, shape=treatment)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  ylim(-45, 20) + xlim(-50, 30)+
  #ylim(-40, 20) + xlim(-50, 30)+
  scale_shape_manual(values=c(21,24), labels = c("Ambient", "OWA"))+
  scale_fill_manual(values=c('#6699CC', "#CC3333"), labels = c("Ambient", "OWA"))+
  guides(shape = guide_legend(override.aes = list(shape = c( 21, 24))))+
  guides(fill = guide_legend(override.aes = list(shape = c( 21, 24))))+
  guides(shape = guide_legend(override.aes = list(size = 5)))+
  theme_bw() +
  theme(legend.position = "none") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 4))+
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank())
F11


# png("PCA_F11.png", res=300, height=5, width=5, units="in")
# ggarrange(F11, nrow = 1, ncol=1)
# dev.off()

ggarrange(F0, F2, F4, F11, nrow = 2, ncol=2)
```



Check on the DE results from the DESeq command way above
```{r}
resultsNames(dds)
#[1] "Intercept"            "generation_F11_vs_F0" "generation_F2_vs_F0"  "generation_F4_vs_F0" 
#[5] "treatment_OA_vs_AM"   "treatment_OW_vs_AM"   "treatment_OWA_vs_AM" 

resAM_OWA <- results(dds, name="treatment_OWA_vs_AM", alpha=0.05)

resAM_OWA <- resAM_OWA[order(resAM_OWA$padj),] #adjusting the p-value to correct for multiple testing
head(resAM_OWA)  
#                                                                      pvalue        padj
#TRINITY_DN22748_c0_g1::TRINITY_DN22748_c0_g1_i4::g.47585::m.47585 1.99509e-10 2.75671e-06
#TRINITY_DN3600_c0_g1::TRINITY_DN3600_c0_g1_i2::g.16079::m.16079   2.18302e-10 2.75671e-06
#TRINITY_DN505_c0_g1::TRINITY_DN505_c0_g1_i6::g.3715::m.3715       5.13460e-10 4.32265e-06
#TRINITY_DN14069_c0_g1::TRINITY_DN14069_c0_g1_i2::g.39418::m.39418 1.29844e-08 6.55866e-05
#TRINITY_DN29_c1_g2::TRINITY_DN29_c1_g2_i14::g.733::m.733          1.17424e-08 6.55866e-05
#TRINITY_DN8532_c1_g1::TRINITY_DN8532_c1_g1_i3::g.30152::m.30152   2.66438e-08 1.12153e-04

summary(resAM_OWA)
#out of 25260 with nonzero total read count
#adjusted p-value < 0.05
#LFC > 0 (up)       : 99, 0.39% few transcripts up or down regualted between OWA and AM
#LFC < 0 (down)     : 75, 0.3%
#outliers [1]       : 4, 0.016%
#low counts [2]     : 0, 0%
#(mean count < 17)

###################################
resAM_OW <- results(dds, name="treatment_OW_vs_AM", alpha=0.05)

resAM_OW <- resAM_OW[order(resAM_OW$padj),]
head(resAM_OW)  
#                                                                        pvalue        padj
#TRINITY_DN29_c1_g2::TRINITY_DN29_c1_g2_i3::g.747::m.747            1.25756e-08 0.000286812
#TRINITY_DN17372_c0_g1::TRINITY_DN17372_c0_g1_i7::g.43175::m.43175  3.11221e-08 0.000354901
#TRINITY_DN29_c1_g2::TRINITY_DN29_c1_g2_i3::g.744::m.744            5.56671e-08 0.000423200
#TRINITY_DN1275_c0_g1::TRINITY_DN1275_c0_g1_i9::g.6875::m.6875      1.45779e-06 0.008311970
#TRINITY_DN17372_c0_g1::TRINITY_DN17372_c0_g1_i11::g.43161::m.43161 2.02842e-06 0.009252455
#TRINITY_DN29_c1_g2::TRINITY_DN29_c1_g2_i14::g.733::m.733           3.11356e-06 0.011442688

summary(resAM_OW)
# out of 25260 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 9, 0.036%
# LFC < 0 (down)     : 8, 0.032%
# outliers [1]       : 4, 0.016%
# low counts [2]     : 2449, 9.7%
# (mean count < 31)


#########################
resAM_OA <- results(dds, name="treatment_OA_vs_AM", alpha=0.05)

resAM_OA <- resAM_OA[order(resAM_OA$padj),] #adjusting the p-value to correct for multiple testing
head(resAM_OA)  
#                                                                      pvalue        padj
# TRINITY_DN2187_c0_g1::TRINITY_DN2187_c0_g1_i2::g.10873::m.10873   4.77741 1.77569e-06 0.00112135
# TRINITY_DN2693_c0_g1::TRINITY_DN2693_c0_g1_i7::g.12416::m.12416   4.90534 9.32642e-07 0.00112135
# TRINITY_DN1347_c0_g1::TRINITY_DN1347_c0_g1_i3::g.7476::m.7476     4.25752 2.06704e-05 0.00815245
# TRINITY_DN21_c1_g1::TRINITY_DN21_c1_g1_i1::g.282::m.282           4.18680 2.82915e-05 0.00815245
# TRINITY_DN6238_c0_g1::TRINITY_DN6238_c0_g1_i1::g.24212::m.24212   4.15680 3.22741e-05 0.00815245
# TRINITY_DN7404_c0_g1::TRINITY_DN7404_c0_g1_i1::g.27475::m.27475   4.08268 4.45191e-05 0.00937128

summary(resAM_OA)
# out of 25260 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 88, 0.35%
# LFC < 0 (down)     : 0, 0%
# outliers [1]       : 4, 0.016%
# low counts [2]     : 23993, 95%
# (mean count < 1796)
```

Plot Individual genes 
```{r}
# Counts of specific top interaction gene! (important validation that the normalization, model is working) 
# OWA vs. AM top gene TRINITY_DN29_c1_g2::TRINITY_DN29_c1_g2_i3::g.747::m.747
d <-plotCounts(dds, gene="TRINITY_DN29_c1_g2::TRINITY_DN29_c1_g2_i3::g.747::m.747", intgroup = (c("treatment","generation")), returnData=TRUE)
d

p <-ggplot(d, aes(x=treatment, y=count, color=treatment, shape=generation)) + 
  theme_minimal() + theme(text = element_text(size=20), panel.grid.major=element_line(colour="grey"))
p <- p + geom_point(position=position_jitter(w=0.2,h=0), size=3)
p <- p + stat_summary(fun = mean, geom = "line")
p <- p + stat_summary(fun = mean, geom = "point", size=5, alpha=0.7) 
p
```



MODEL NUMBER 2 - subset to focus on effect of treatment for each generation
```{r}

dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData=conds, 
                              design= ~ treatment)

dim(dds)
# [1] 67916     38

# Filter 
dds <- dds[rowSums(counts(dds) >= 15) >= 28,] # 25260
nrow(dds) 

# Subset the DESeqDataSet to the specific level of the "generation" factor
dds_sub <- subset(dds, select = generation == 'F0')
dim(dds_sub) # [1] 25260    12

# Perform DESeq2 analysis on the subset
dds_sub <- DESeq(dds_sub)

resultsNames(dds_sub)
# [1] "Intercept"  "treatment_OA_vs_AM"  "treatment_OW_vs_AM"  "treatment_OWA_vs_AM"


res_F0_OWvAM <- results(dds_sub, name="treatment_OW_vs_AM", alpha=0.05)

res_F0_OWvAM <- res_F0_OWvAM[order(res_F0_OWvAM$padj),]
head(res_F0_OWvAM)
# log2 fold change (MLE): treatment OW vs AM 
# Wald test p-value: treatment OW vs AM 
# DataFrame with 6 rows and 6 columns
#                                                                    baseMean log2FoldChange     lfcSE      stat      pvalue        padj
# TRINITY_DN30_c0_g2::TRINITY_DN30_c0_g2_i1::g.130::m.130           32956.030        3.74319  0.292239   12.8086 1.46698e-37 3.63048e-33
# TRINITY_DN33_c0_g1::TRINITY_DN33_c0_g1_i18::g.235::m.235           4622.325        3.72565  0.296419   12.5689 3.13154e-36 3.87496e-32
# TRINITY_DN9365_c0_g1::TRINITY_DN9365_c0_g1_i1::g.32056::m.32056    3814.681        3.62002  0.292971   12.3562 4.50893e-35 3.71957e-31
# TRINITY_DN258_c0_g1::TRINITY_DN258_c0_g1_i18::g.1704::m.1704       4670.245        3.72840  0.317074   11.7588 6.36643e-32 3.20614e-28
# TRINITY_DN258_c0_g1::TRINITY_DN258_c0_g1_i13::g.1700::m.1700       5937.496        3.30257  0.280896   11.7573 6.47758e-32 3.20614e-28
# TRINITY_DN10209_c0_g2::TRINITY_DN10209_c0_g2_i3::g.33784::m.33784   279.313        3.00313  0.264004   11.3753 5.55141e-30 2.28977e-26

summary(res_F0_OWvAM)
# out of 25260 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 3468, 14%
# LFC < 0 (down)     : 2049, 8.1%
# outliers [1]       : 22, 0.087%
# low counts [2]     : 490, 1.9%
# (mean count < 20)


### Plot Individual genes ### 

# Counts of specific top interaction gene! (important validation that the normalization, model is working)
# OW vs. AM top gene TRINITY_DN30_c0_g2::TRINITY_DN30_c0_g2_i1::g.130::m.130
d <-plotCounts(dds_sub, gene="TRINITY_DN30_c0_g2::TRINITY_DN30_c0_g2_i1::g.130::m.130", intgroup = (c("treatment","generation")), returnData=TRUE)
d

p <-ggplot(d, aes(x=treatment, y=count, color=treatment, shape=generation)) + 
  theme_minimal() + theme(text = element_text(size=20), panel.grid.major=element_line(colour="grey"))
p <- p + geom_point(position=position_jitter(w=0.2,h=0), size=3)
p <- p + stat_summary(fun = mean, geom = "point", size=5, alpha=0.7) 
p


### We can make an MA plot (sideways volcano) ###
plotMA(res_F0_OWvAM, ylim=c(-6,6))
```

Heatmap of top 20 genes sorted by pvalue
```{r}
library(pheatmap)

# By environment
vsd <- vst(dds_sub, blind=FALSE)

topgenes <- head(rownames(res_F0_OWvAM),20)
mat <- assay(vsd)[topgenes,]
mat <- mat - rowMeans(mat) #scaling matrix so it's normalized for comparison (subtracting means)
df <- as.data.frame(colData(dds_sub)[,c("generation","treatment")])
pheatmap(mat, annotation_col=df)
pheatmap(mat, annotation_col=df, cluster_cols = F)
```
PLOT OVERLAPPING DEGS IN VENN EULER DIAGRAM
```{r}

# Since we have three contrasts, let’s make a Venn (or Euler) Diagram to see how similar or different the DGE is from Ambient for OA, OW, and OWA at F0!

# making lists of differentially expressed genes For OW vs AM
res_F0_OWvAM <- results(dds_sub, name="treatment_OW_vs_AM", alpha=0.05) # currently only using data for F0 generation (dds_sub), could also filter based on test "stat" or log2 fold change to split between up and down regulated gene sets
res_F0_OWvAM <- res_F0_OWvAM[order(res_F0_OWvAM$padj),]
head(res_F0_OWvAM)

summary(res_F0_OWvAM)
res_F0_OWvAM <- res_F0_OWvAM[!is.na(res_F0_OWvAM$padj),] # removes samples with no call
degs_F0_OWvAM <- row.names(res_F0_OWvAM[res_F0_OWvAM$padj < 0.05,])

### For OA vs AM ###
res_F0_OAvAM <- results(dds_sub, name="treatment_OA_vs_AM", alpha=0.05)
res_F0_OAvAM <- res_F0_OAvAM[order(res_F0_OAvAM$padj),]
head(res_F0_OAvAM)

summary(res_F0_OAvAM)
res_F0_OAvAM <- res_F0_OAvAM[!is.na(res_F0_OAvAM$padj),]
degs_F0_OAvAM <- row.names(res_F0_OAvAM[res_F0_OAvAM$padj < 0.05,])

### For OWA vs AM ###
res_F0_OWAvAM <- results(dds_sub, name="treatment_OWA_vs_AM", alpha=0.05)
res_F0_OWAvAM <- res_F0_OWAvAM[order(res_F0_OWAvAM$padj),]
head(res_F0_OWAvAM)

summary(res_F0_OWAvAM)
res_F0_OWAvAM <- res_F0_OWAvAM[!is.na(res_F0_OWAvAM$padj),]
degs_F0_OWAvAM <- row.names(res_F0_OWAvAM[res_F0_OWAvAM$padj < 0.05,])

library(eulerr)

# Total number of genes
length(degs_F0_OAvAM)  # 602
length(degs_F0_OWvAM)  # 5517 
length(degs_F0_OWAvAM)  # 3918

# Intersections
length(intersect(degs_F0_OAvAM,degs_F0_OWvAM))  # 444
length(intersect(degs_F0_OAvAM,degs_F0_OWAvAM))  # 380
length(intersect(degs_F0_OWAvAM,degs_F0_OWvAM))  # 2743

intWA <- intersect(degs_F0_OAvAM,degs_F0_OWvAM)
length(intersect(degs_F0_OWAvAM,intWA)) # 338 overlap between all 3 conditions

# Number unique
602-444-380+338 # 116 OA
5517-444-2743+338 # 2668 OW 
3918-380-2743+338 # 1133 OWA

444-338 # 106 OA & OW
380-338 # 42 OA & OWA
2743-338 # 2405 OWA & OW


# Note that the names are important and have to be specific to line up the diagram
fit1 <- euler(c("OA" = 116, "OW" = 2668, "OWA" = 1133, "OA&OW" = 106, "OA&OWA" = 42, "OW&OWA" = 2405, "OA&OW&OWA" = 338))


plot(fit1,  lty = 1:3, quantities = TRUE)
# lty changes the lines

# different diagram aesthetic
plot(fit1, quantities = TRUE, fill = "transparent",
     lty = 1:3,
     labels = list(font = 4))


#cross check
2668+2405+338+106 # 5517, total OW
1133+2405+338+42 # 3918, total OWA
116+42+106+338    # 602, total OA
```

